core/unit_test/tstParticleGridCommunication.hpp

```diff
@@ -0,0 +1,445 @@
+            EXPECT_LE( pos_host( i, Cajita::Dim::K ), hi_z );
+            EXPECT_GE( pos_host( i, Cajita::Dim::I ), lo_x );
+            EXPECT_GE( pos_host( i, Cajita::Dim::J ), lo_y );
+            EXPECT_GE( pos_host( i, Cajita::Dim::K ), lo_z );
+        }
+    }
+
+    // Make sure the particles are still unique.
+    int final_total = 0;
+    int initial_total = 0;
+    MPI_Allreduce( &new_num_data, &final_total, 1, MPI_INT, MPI_SUM,
+                   MPI_COMM_WORLD );
+    MPI_Allreduce( &num_data, &initial_total, 1, MPI_INT, MPI_SUM,
+                   MPI_COMM_WORLD );
+    EXPECT_EQ( final_total, initial_total );
+}
+
+//---------------------------------------------------------------------------//
+void testHalo( const int halo_width, const int test_type )
+{
+    // Create the MPI partitions.
+    Cajita::UniformDimPartitioner partitioner;
+
+    // Create the global MPI decomposition mesh.
+    std::array<double, 3> lo_corner = { 0, 0, 0 };
+    std::array<double, 3> hi_corner = { 1, 1, 1 };
+    std::array<int, 3> num_cell = { 10, 10, 10 };
+    auto global_mesh =
+        Cajita::createUniformGlobalMesh( lo_corner, hi_corner, num_cell );
+
+    // Create the global grid.
+    std::array<bool, 3> is_periodic = { true, true, true };
+    auto global_grid = Cajita::createGlobalGrid( MPI_COMM_WORLD, global_mesh,
+                                                 is_periodic, partitioner );
+
+    // Create a  grid local_grid.
+    auto local_grid = Cajita::createLocalGrid( global_grid, 4 );
+    auto local_mesh = Cajita::createLocalMesh<TEST_DEVICE>( *local_grid );
+
+    // Make some data to migrate, one for each rank neighbor. This tests
+    // only for neighbors within one ghost shell.
+    int num_data = 27;
+    const int id_index = 0;
+    const int pos_index = 1;
+    using DataTypes = Cabana::MemberTypes<int, double[3]>;
+    Cabana::AoSoA<DataTypes, Kokkos::HostSpace> data_host( "host", num_data );
+
+    // Get mesh info.
+    auto dx = local_grid->globalGrid().globalMesh().cellSize( Cajita::Dim::I );
+    auto dy = local_grid->globalGrid().globalMesh().cellSize( Cajita::Dim::J );
+    auto dz = local_grid->globalGrid().globalMesh().cellSize( Cajita::Dim::K );
+    auto lo_x = local_mesh.lowCorner( Cajita::Own(), Cajita::Dim::I );
+    auto lo_y = local_mesh.lowCorner( Cajita::Own(), Cajita::Dim::J );
+    auto lo_z = local_mesh.lowCorner( Cajita::Own(), Cajita::Dim::K );
+    auto width_x = local_mesh.extent( Cajita::Own(), Cajita::Dim::I ) / 2.0;
+    auto width_y = local_mesh.extent( Cajita::Own(), Cajita::Dim::J ) / 2.0;
+    auto width_z = local_mesh.extent( Cajita::Own(), Cajita::Dim::K ) / 2.0;
+    auto center_x = width_x + lo_x;
+    auto center_y = width_y + lo_y;
+    auto center_z = width_z + lo_z;
+    auto shift_x = width_x - ( halo_width - 0.1 ) * dx;
+    auto shift_y = width_y - ( halo_width - 0.1 ) * dy;
+    auto shift_z = width_z - ( halo_width - 0.1 ) * dz;
```

Comment:
All of the setup code looks the same for both tests - perhaps it can be combined.

Reply:
Combined